### ISSUE TITLE
Showing almanac, ephemeris and other sat data – please discuss

### ISSUE THREAD

**Issue by mvglasow** [mvglasow/satstat#69]
This ticket came out of merge request #48, which already got some input.

Currently, the sky plot and SNR graph use two different colors to show whether or not a satellite is used in the current fix; signal-to-noise ratio (SNR) is represented by dot diameter/bar length. The SNR chart additionally shows which GNSS each satellite belongs to (GPS, GLONASS, Galileo, others) – though this is going to be improved, see #64.

The merge request proposes to also implement visualization of almanac and ephemeris availability. In total, each satellite has five "dimensions" which could be of interest:

* whether the satellite is used in the current fix (currently represented by color)
* Signal-to-noise ratio (currently represented by size)
* GNSS system (currently represented by position in SNR chart, possibly by shape in sky plot)
* almanac availability
* ephemeris availability

Availability of augmentation systems (SBAS) has also been suggested, but I'd have to look into how this is implemented: is that a per-satellite flag, a per-GNSS flag or a global one? These systems present themselves as additional satellites (see https://github.com/mvglasow/satstat/wiki/NMEA-IDs#sbas), which should already be visible.

I'm looking for ideas how to best represent all of these. Some devices report bogus data (e.g. a satellite is used in the fix despite having no almanac or ephemeris), which also need to be visualized. Therefore, we really have to treat these five dimensions independently.

We may need to limit ourselves to showing certain dimensions only in one of the two graphs.

As for visual dimensions, apart from size, color and position, we could also consider shape (circle, square, triangle...), luminance and saturation, as well as border color, or maybe hollow vs. filled dots.

We'd have to watch out for "interference" between dimensions: fill/border distinctions as well as different shapes get hard to distinguish at small sizes (weak signal or lots of sats in the SNR chart); using luminance and saturation together might result in very similar colors.

Suggestions are welcome!

**Comment by andree182** [comment]
Hi! I actually part of this already implemented, for quite some time even :)) But I have yet to merge the  recent changes you did, and cleanup - then I'd like to send you the pull request. But in the meantime you can try to build and check it. Basically:

* satellite type - color
* satellita usage - more dim color
* almanac/ephemeris - shown by circle around the satellite (seems it needs some work on high DPI :) )

**Comment by andree182** [comment]
In my implementation (now largerly obsolete), I did a similar thing...

* different colors for different "networks", network names to be shown under 
* SNR represented as size of the dot
* unused satellites drawn with less saturation
* almanac/ephemeris as two half-circles (it's not that useful and reliable info anyhow)

this way, you don't need different shapes, yet with 2-3-4 networks, the screen is still perfectly "readable".

I'll try to find the yet-useful commits and send the pull request in other thread...

**Comment by mvglasow** [comment]
In the meantime, I've added the GNSS name to the SNR graph as a label underneath the bars. Currently I'm developing in the `satstat-mapsforge0.6.1` branch as I'm planning a mapsforge upgrade in the next version – that branch also has some changes which are not related to mapsforge.

The two half-circles are an interesting approach – hadn't thought of that possibility yet.

As for reliability of almanac and ephemeris – that is one part of what SatStat is about: visualizing location provider and sensor data, including bogus data returned by the device. Therefore, visibility of that information is just as important as the rest.

I don't want to stop your enthusiasm, but I'd like to collect some more input before starting to implement things. I don't want to disappoint anyone by having them submit a pull request which ends up being obsoleted by discussion results. (Of course, if you are working on this for your own use anyway, feel free to do so.) However, if you have a working version (can be based off an older SatStat build) and you can post a screenshot, I'd be interested in seeing that.

**Comment by andree182** [comment]
Here's what I had:

![img_20160515_103459](https://cloud.githubusercontent.com/assets/1175689/15272949/c53b3bd8-1a89-11e6-9082-33030324281d.jpg)
![screenshot_20160515-103116](https://cloud.githubusercontent.com/assets/1175689/15272950/c86ace72-1a89-11e6-83c1-2fe06448061a.png)

I changed the device in the meantime, there seem to be no ephemeris/almanac or something, so no demo of that. But at least, you can see the colors, a "leveler bubble", and sunlight map (may not be that interesting for this app?).

Don't worry about the enthusiasm - if I mourned every code that got lost, I'd get nowhere... :)

**Comment by mvglasow** [comment]
About the circles, I just saw a screenshot of another app that made me rethink my stance on that. My previous concern had been that drawing a ring inside the dot would be hard to detect with very small dots. Drawing it around the dot, as a kind of "halo", however, would not suffer from this as long as we ensure a certain minimum stroke width and/or sufficient contrast.

That way, we could represent the information as follows:

* GNSS system: color
* Satellite used in fix: dimmed vs. full brightness
* SNR: size
* Almanac/ephemeris: halo (or background in bar chart)

To distinguish between almanac and ephemeris, we could use a combination of hue and brightness. Ephemeris is expressed by brightness while almanac is expressed by a hue:

* A satellite with neither almanac nor ephemeris has no halo.
* A satellite with ephemeris but no almanac has a gray halo.
* A satellite with an almanac but no ephemeris has a dark colored halo.
* A satellite with both almanac and ephemeris has a bright colored halo.

**Comment by gdt** [comment]
I have some contrarian thoughts.   First, I think it would be good to make changes incrementally to see what's really useful for people, and to understand what kind of information is actually reported.  In particular, I think representing almanac is less important than it seems.

Second, this is a tricky user interface issue, basically trying to get too much information into someone's brain.  Besides the basic overload issue, there's the issue that there are different users with different levels of concern and understanding.  I think we could easily end up with clutter/confusing.

My first suggestion is that the first step, rather than changing the graph page, should be to add a new pane that has a text table of constellation/id/alamanac/ephemeris/used, and to try to put whatever information is available (start valid time?  end of valid time?  probably just end) in there.   That will enable the nerdy to understand (and fulfill the debugging role), and more importantly it will give us feedback on what is going to be useful to represent.

My second suggestion is to try to represent almanac/ephemeris in the bars, not the circles, as the first step.   One reason is that this is how Garmin does it, and it's been very useful/reasonable interface for me for 20 years.  The second is that the circles are in danger of becoming cluttered, as it is already hard to separate satellites some times.  I can send a screenshot if that helps, but the bars are hollow without ephemeris and solid with.

I also note that if there is no almanac, non-tracked satellites will not be present in the sky view.  That is perhaps enough of a clue in the main page.

In the bars, I think they should be hollow (just the two sides and top) if there is no ephemeris and the satellite is not in the solution.   Then the bottom half can be filled with color (as it is now) if there is ephemeris, and the top half if it's in the solution.   If there's a real desire to show almanac, I would have no mark for almanac present, and a black slash across the bar for missing almanac, or perhaps a black slash to the max height, even for satellites not being tracked.  The no-almanac display will be glaring and ugly, but it's a very bad condition for a receiver, and it's temporary.   (I suspect I had it for the first 20 minutes of turning on my phone out of the box and haven't had it again since.)

As a next step, we could add a "D" in the bar to satellites that have differential data, and some notation in the fix info if the fix is "3D diff" vs "3D".  Again this is how Garmin does it (except that they only put D in bars if the fix is differential, I think).

After that is done, we could reassess if adding things to the circles would help.  My guess is that there won't be much desire for that once the bars show status.

As for your dot plan, I suggest that you remove almanac from the mix or deemphasize it, at least until we have some evidence that it's actually useful.   For GPS, almanac is pretty much all or nothing, and receivers almost always have it.   Having a details pane avoids the residual need to show it, in my view.   Then, you could use three levels of dim: dimmest for no ephemeris, mid for ephemeris not in fix, and bright for ephemeris and in fix.   (I suspect ephemeris not in fix is very rare.)   And if you want to show almanac, you can put a black slash across it if missing.  Note that the only way to know where a satellite is w/o almanac is to have ephemeris for that satellite.

**Comment by gdt** [comment]
While I've been pushing hollow bars, the dim bar approach to ephemeris seems totally reasonable too.

**Comment by gdt** [comment]
I am 95% sure that in GPS, the almanac message (transmitted by each satellite) takes 15 minutes and once received provides almanac data for every satellite in the constellation.  While in theory the control segment could have different almanac messages on different satellites and not have full coverage, I've never heard of that.

Can someone explain almanac for GLONASS?   Is it the same "you have it all or you don't have any" notion?   My Etrex 30 seems to behave that way.

**Comment by mvglasow** [comment]
@gdt Valid points.

> this is a tricky user interface issue, basically trying to get too much information into someone's brain.

Agreed – if this were easy, I'd have merged #48 long ago and we wouldn't be having this discussion.

> a text table of constellation/id/alamanac/ephemeris/used, and to try to put whatever information is available (start valid time? end of valid time? probably just end) in there

Off the top of my head, I don't think the Android API provides a lot of data besides flags for availability. But I'd have to check on that to be sure.

> there are different users with different levels of concern and understanding

Indeed, different users will have different needs and preferences. Maybe it's best to just add a preference so users can choose if and where they want almanac/ephemeris visualized?

> I also note that if there is no almanac, non-tracked satellites will not be present in the sky view

If the vendor has done their homework, that is. A lot of vendors don't, though – that goes for internals of GPS as well as for GSM/UMTS/LTE cells, and one use case for SatStat is to expose that.

> As a next step, we could add a "D" in the bar to satellites that have differential data

Unfortunately, nothing in Android's Location API exposes that kind of information. At the most, there's raw NMEA data, which one could look into, but a bunch of GPS chips in today's smartphones don't even support NMEA but speak some vendor protocol, with a vendor-specific driver translating that into location information.

> While in theory the control segment could have different almanac messages on different satellites and not have full coverage, I've never heard of that.

The Android API indeed has `GpsSatellite#hasAlmanac()`, which would allow just that. By exposing that without too much translation in between, we'll find out.

**Comment by gdt** [comment]
That's interesting about the API not having D and reporting almanac per-satellite.   I think that's all the more reason to have a pane with a nerdy text table of all information available.    That will let us understand what's out there, and it will let people who need to debug their ROM really know precisely what's going on.

With respect to non-alamnac and tracking I meant that if there is no almanac and no ephemeris, there is no way to know the position.  I agree that a ROM could report bogus positions in that case.   But I think it's important to keep two goals in mind and slightly separated:
  * showing the user the state of a GPS receiver that is operating sanely, finding out about the quality of the current navigation solution
  * showing the user the gory details of a possibly-misbehaving system including things that shouldn't be true

Basically I am concerned about compromising the first in the pursuit of the second.    I think 95%+ users will want the first only, and of the remaining 5% like us, we'll want the first view 99% of the time.   That leads me to show what's normal, and to put the exceptional stuff as a positive graphical indication, missing when it's normal.

Thanks for listening; I know I'm ranting a lot.

**Comment by andree182** [comment]
Btw, @mvglasow, you may find this useful - https://github.com/mvglasow/satstat/commit/ecea8b5ed48344c660d27c4ebda16035690d5ace (the initializeGrid() and getGridPos() cleanup). But if you won't need it immediately while implementing this ticket, I'll rework the patch and create pull request later.
